src/System.Text.Utf8/src/System/Text/Utf8/Utf8String.cs

```diff
@@ -24,61 +19,36 @@ public partial struct Utf8String : IEnumerable<Utf8CodeUnit>, IEquatable<Utf8Str
 
         // TODO: Validate constructors, When should we copy? When should we just use the underlying array?
         // TODO: Should we be immutable/readonly?
-        public Utf8String(ByteSpan buffer)
+        public Utf8String(Span<byte> buffer)
         {
             _buffer = buffer;
-            _bytes = null;
-            _index = 0;
-            _length = 0;
         }
 
-        public Utf8String(byte[] utf8bytes) : this(utf8bytes, 0, utf8bytes.Length)
+        public Utf8String(byte[] utf8bytes)
         {
+            _buffer = new Span<byte>(utf8bytes);
         }
 
         public Utf8String(byte[] utf8bytes, int index, int length)
         {
-            if (utf8bytes == null)
-            {
-                throw new ArgumentNullException("utf8bytes");
-            }
-            if (index + length > utf8bytes.Length)
-            {
-                throw new ArgumentOutOfRangeException("index");
-            }
-
-            _buffer = default(ByteSpan);
-            _bytes = utf8bytes;
-            _index = index;
-            _length = length;
+            _buffer = new Span<byte>(utf8bytes, index, length);
         }
 
         // TODO: reevaluate implementation
         public Utf8String(IEnumerable<UnicodeCodePoint> codePoints)
         {
             int len = GetUtf8LengthInBytes(codePoints);
-            byte[] utf8bytes = new byte[len];
-            unsafe
+            _buffer = new Span<byte>(new byte[len]);
+            Span<byte> span = _buffer;
+            foreach (UnicodeCodePoint codePoint in codePoints)
```

Comment:
I understand that. But it looks from the code that the only reason why you write to the span is to validate the CodePoints, no? And if that's indeed true, than you are in fact, just validating it, even though you write bytes in order to do so. Let me know if you misunderstood something here.